src/server/multi_command_squasher.cc

```diff
@@ -66,6 +67,42 @@ MultiCommandSquasher::ShardExecInfo& MultiCommandSquasher::PrepareShardInfo(Shar
   return sinfo;
 }
 
+// Atomic transactions (that have all keys locked) perform hops and run squashed commands via
+// stubs, non-atomic ones just run the commands in parallel (hopefully running many hops inline).
+void MultiCommandSquasher::PerformHop() {
+  auto* tx = cntx_->transaction;
+
+  // Simple fanout for non-atomic.
+  if (!IsAtomic()) {
+    shard_set->RunBlockingInParallel([this, tx](auto* es) { SquashedHopCb(tx, es); },
+                                     [this](auto sid) { return !sharded_[sid].cmds.empty(); });
+    return;
+  }
+
+  cntx_->cid = base_cid_;
+  auto run_cb = absl::bind_front(&MultiCommandSquasher::SquashedHopCb, this);
+
+  // If all commands fit into a single batch, run them as a real single hop without multi overhead.
+  // Doesn't work with replication, so disallow if inline scheduling is not allowed.
+  bool singlehop_possible = IsAtomic() && !tx->IsScheduled() && cmds_.empty();
+  if (singlehop_possible && ServerState::tlocal()->AllowInlineScheduling()) {
```

Comment:
this is not safe obviously... making this stuff work with replication is a little cumbersome